File: python/app/controllers/ai_controller.py
from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.database import get_db
from app.models.transaction import Transaction
from app.services import ai_service
from app.middleware.auth import get_current_user_id

router = APIRouter(prefix="/api/ai", tags=["ai"])


class ClassifyRequest(BaseModel):
    description: str


class AnalyzeRequest(BaseModel):
    transactions: list[dict] | None = None


class ChatRequest(BaseModel):
    messages: list[dict]


@router.post("/classify")
async def classify(req: ClassifyRequest):
    result = await ai_service.classify(req.description)
    return result


@router.post("/analyze-spending")
async def analyze_spending(
    req: AnalyzeRequest,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
):
    transactions_data = req.transactions
    if transactions_data is None:
        txns = (
            db.query(Transaction)
            .filter(Transaction.user_id == user_id)
            .order_by(Transaction.transaction_date.desc())
            .limit(100)
            .all()
        )
        transactions_data = [t.to_dict() for t in txns]

    result = await ai_service.analyze_spending(transactions_data)
    return result


@router.post("/detect-anomalies")
async def detect_anomalies(
    req: AnalyzeRequest,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
):
    transactions_data = req.transactions
    if transactions_data is None:
        txns = (
            db.query(Transaction)
            .filter(Transaction.user_id == user_id)
            .order_by(Transaction.transaction_date.desc())
            .limit(100)
            .all()
        )
        transactions_data = [t.to_dict() for t in txns]

    result = await ai_service.detect_anomalies(transactions_data)
    return result


@router.post("/predict-budget")
async def predict_budget(
    req: AnalyzeRequest,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
):
    transactions_data = req.transactions
    if transactions_data is None:
        txns = (
            db.query(Transaction)
            .filter(Transaction.user_id == user_id)
            .order_by(Transaction.transaction_date.desc())
            .limit(100)
            .all()
        )
        transactions_data = [t.to_dict() for t in txns]

    result = await ai_service.predict_budget(transactions_data)
    return result


@router.post("/tips")
async def tips(
    req: AnalyzeRequest,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
):
    transactions_data = req.transactions
    if transactions_data is None:
        txns = (
            db.query(Transaction)
            .filter(Transaction.user_id == user_id)
            .order_by(Transaction.transaction_date.desc())
            .limit(100)
            .all()
        )
        transactions_data = [t.to_dict() for t in txns]

    result = await ai_service.get_tips(transactions_data)
    return result


@router.post("/chat")
async def chat(req: ChatRequest):
    result = await ai_service.chat(req.messages)
    return result

File: node/src/controllers/aiController.js
const { Transaction, Category } = require('../models');
const aiService = require('../services/aiService');
const { Op } = require('sequelize');

exports.classify = async (req, res, next) => {
  try {
    const { description } = req.body;
    if (!description) {
      return res.status(400).json({ error: true, message: 'Description is required' });
    }

    const categories = await Category.findAll({ where: { userId: req.userId } });
    const result = await aiService.classifyTransaction(description, categories);

    res.json({ suggestedCategory: result, description });
  } catch (err) {
    next(err);
  }
};

exports.analyzeSpending = async (req, res, next) => {
  try {
    const { query } = req.body;
    const transactions = await Transaction.findAll({
      where: { userId: req.userId },
      order: [['transactionDate', 'DESC']],
      limit: 100,
    });

    const result = await aiService.analyzeSpending(transactions, query);
    res.json({ analysis: result, transactionsCount: transactions.length });
  } catch (err) {
    next(err);
  }
};

exports.detectAnomalies = async (req, res, next) => {
  try {
    const transactions = await Transaction.findAll({
      where: { userId: req.userId },
      order: [['transactionDate', 'DESC']],
      limit: 200,
    });

    const result = await aiService.detectAnomalies(transactions);
    res.json(result);
  } catch (err) {
    next(err);
  }
};

exports.predictBudget = async (req, res, next) => {
  try {
    const transactions = await Transaction.findAll({
      where: {
        userId: req.userId,
        transactionDate: {
          [Op.gte]: new Date(new Date().setFullYear(new Date().getFullYear() - 1)),
        },
      },
      order: [['transactionDate', 'ASC']],
    });

    const result = await aiService.predictBudget(transactions);
    res.json(result);
  } catch (err) {
    next(err);
  }
};

exports.tips = async (req, res, next) => {
  try {
    const transactions = await Transaction.findAll({
      where: { userId: req.userId },
      order: [['transactionDate', 'DESC']],
      limit: 100,
    });

    const result = await aiService.getFinancialTips(transactions);
    res.json(result);
  } catch (err) {
    next(err);
  }
};

exports.chat = async (req, res, next) => {
  try {
    const { message, context } = req.body;

    if (!message) {
      return res.status(400).json({ error: true, message: 'Message is required' });
    }

    const result = await aiService.chat(message, context || []);
    res.json(result);
  } catch (err) {
    next(err);
  }
};
